enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+
+For every unique DNS name used in EgressFirewall rules, OVN-K cluster manager will create a corresponding `DNSNameResolver` CR.
+The name of the CR will be assigned using a hash function (similar to the ComputeHash
+[here](https://github.com/openshift/kubernetes/blob/master/pkg/controller/controller_utils.go#L1157-L1172))
+prefixed by `dns-`. The input to the hash function will be the DNS name. It will also delete a `DNSNameResolver` CR, when all the rules containing
+the corresponding DNS name are deleted.
+
+#### OVN-K master(s)
+
+Add the support of a new flag `--enable-dns-name-resolver` to configure the use of `DNSNameResolver` CRD with OVN-K.
+
+The OVN-K master(s) will watch the `DNSNameResolver` CRs. Whenever the `.status` of the CRs will be updated with new IP addresses and corresponding
+TTL information for a DNS name, OVN-K master(s) will update the `AddressSet` mapped to the DNS name. This `AddressSet` will be linked
+to the ACL rule(s) for the EgressFirewall rule(s) in which the DNS name is used. This will ensure that the latest IP addresses are always updated in
+the `AddressSets`.
+
+#### Cluster Network Operator
+
+To add the support of `DNSNameResolver` CRD in OVN-K, add the flag `--enable-dns-name-resolver` to the corresponding OVN-K pods.
+
+
+### Risks and Mitigations
+
+* The `DNSNameResolver` CR will be created by OVN-K Cluster manager whenever a new DNS name is used in a EgressFirewall rule. The CR will be deleted
+when the corresponding DNS name is not used in any of the EgressFirewall rules. The `DNSNameResolver` CR should not be modified (created or deleted or
+updated) by a user. Doing so may lead to undesired behavior of EgressFirewall.
+* During the upgrade process if the OVN-K cluster manager and master pods are upgraded first and then the Cluster DNS operator and the CoreDNS pods, the
+EgressFirewall functionality may break. The details can be found in [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy) section
+
+### Drawbacks
+
+* Whenever there's a change in the IP addresses or the next lookup time (TTL + last lookup time) for a DNS name, the additional step of updating the
+related `DNSNameResolver` CRs will be executed. This will add some delay to the DNS lookup process. However, this will only
+happen whenever there's a change in the DNS information.
+* For wildcard DNS names, it is not possible to pre-emptively find out the matching regular DNS names which have DNS records in the upstream nameservers.
+For these regular DNS names the first lookup may originate from a workload pod, in which case the pod may
+see an initial delay until the underlying ACL rules are updated with the corresponding IP addresses. However, this will happen only for the first DNS lookup
+of a particular DNS name by any pod. Subsequently, it will be handled by the `DNSNameResolver` controller to keep the IP addresses updated.
+
+
+## Design Details
+
+### Open Questions [optional]
```

Comment:
We need to find out how this impacts DNS lookups, and generally how this works, when there are multiple Egress Firewalls, possibly with overlapping wildcard DNS names.